### PR TITLE
Use backoff delay for New-AzureRmResourceGroupDeployment query calls

### PR DESCRIPTION
**Pull request by TianoMS**
This PR fixes two issues:
https://github.com/Azure/azure-powershell/issues/1808,
https://github.com/Azure/azure-powershell/issues/358

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@TianoMS__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <p>
        It looks like you're working at Microsoft (tiano). If you're full-time, we DON'T require a contribution license agreement.
    </p>
    <p>
        If you are a vendor, DO please sign the electronic contribution license agreement. It will take 2 minutes and there's no faxing! https://cla.azure.com.
    </p>

TTYL, AZPRBOT;

**Comment by TianoMS** [comment]
On demand job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/764/

File: src/ResourceManager/Resources/Commands.Resources/Models.ResourceGroups/ResourceClient.cs [hunk 1]
```diff
@@ -325,16 +325,17 @@ public static List<string> ParseDetailErrorMessage(string statusMessage)
             params string[] status)
         {
             DeploymentExtended deployment;
+            Delay stepedDelay = new Delay(firstDelay: 0);
 
             do
             {
+                TestMockSupport.Delay(stepedDelay.GetNextDelay());
```
Comment (vivsriaus) [inline src/ResourceManager/Resources/Commands.Resources/Models.ResourceGroups/ResourceClient.cs]: Why can't we just do:
int counter = 0;
do
{
  counter =  counter + 5000 > 60000 ? 60000 : counter + 5000;
  TestMockSupport.Delay(counter);
...
}

Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.Resources/Models.ResourceGroups/ResourceClient.cs]: Per our sync up, will change it to the simple way as we don't see future reuse of this.

File: src/Common/Commands.Common/TestMockSupport.cs [hunk 2]
```diff
@@ -47,4 +47,37 @@ public static void Delay(TimeSpan duration)
...
+
+        private int maxDelay;
+
+        private int nextDelay;
+
+        public Delay(int backoffStep = DefaultBackoffStepInMilliseconds, int firstDelay = DefaultDelayInMilliseconds, int maxDelay = DefaultMaxDelayInMilliseconds)
+        {
+            this.backoffStep = backoffStep;
+            this.maxDelay = maxDelay;
+
```
Comment (gandhiniraj) [inline src/Common/Commands.Common/TestMockSupport.cs]: nit: unwanted new line

File: src/Common/Commands.Common/TestMockSupport.cs [hunk 3]
```diff
@@ -47,4 +47,37 @@ public static void Delay(TimeSpan duration)
             }
         }
     }
+
+    public class Delay
+    {
+        public const int DefaultMaxDelayInMilliseconds = 60000;
+
+        public const int DefaultBackoffStepInMilliseconds = 5000;
```
Comment (gandhiniraj) [inline src/Common/Commands.Common/TestMockSupport.cs]: Can we please add comment to explain the use?

File: src/Common/Commands.Common/TestMockSupport.cs [hunk 4]
```diff
@@ -47,4 +47,37 @@ public static void Delay(TimeSpan duration)
...
+
+        public const int DefaultDelayInMilliseconds = 0;
```
Comment (gandhiniraj) [inline src/Common/Commands.Common/TestMockSupport.cs]: Same here

File: src/ResourceManager/Resources/Commands.Resources/Models.ResourceGroups/ResourceClient.cs [hunk 5]
```diff
@@ -325,16 +325,18 @@ public static List<string> ParseDetailErrorMessage(string statusMessage)
             params string[] status)
         {
             DeploymentExtended deployment;
+            Delay stepedDelay = new Delay(firstDelay: 0);
```
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.Resources/Models.ResourceGroups/ResourceClient.cs]: nit: stepped

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/NewAzureResourceCmdlet.cs [hunk 6]
```diff
@@ -63,9 +63,10 @@ public sealed class NewAzureResourceCmdlet : ResourceManipulationCmdletBase
         /// <summary>  
         /// Gets or sets the plan object.  
         /// </summary>  
+        [Alias("SkuObject")]  
         [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "A hash table which represents sku properties.")]  
         [ValidateNotNullOrEmpty]  
-        public Hashtable SkuObject { get; set; }  
+        public Hashtable Sku { get; set; }
```
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/NewAzureResourceCmdlet.cs]: We should also update help with an example with this parameter name change and add unit test.

File: src/Common/Commands.Common/TestMockSupport.cs (hunk 2 above)
Comment (TianoMS) [inline src/Common/Commands.Common/TestMockSupport.cs]: Talked with Vivek yesterday. We don't see future reuse of this delay pattern, so I changed it to the simple way to get incremental stepped delay. I committed another change and this class was removed.

**Comment by TianoMS** [comment]
On demand job: http://azuresdkci.cloudapp.net/view/1-AzurePowerShell/job/powershell-demand/768/

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/NewAzureResourceCmdlet.cs [hunk 7]
```diff
@@ -56,17 +56,18 @@ public sealed class NewAzureResourceCmdlet : ResourceManipulationCmdletBase
         /// <summary>
         /// Gets or sets the plan object.
         /// </summary>
+        [Alias("PlanObject")]
         [Parameter(Mandatory = false, HelpMessage = "A hash table which represents resource plan properties.")]
         [ValidateNotNullOrEmpty]
-        public Hashtable PlanObject { get; set; }
+        public Hashtable Plan { get; set; }
```
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/NewAzureResourceCmdlet.cs]: Here we should add a help which suggest aliasing will still work

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/NewAzureResourceCmdlet.cs [hunk 8]
```diff
@@ -56,17 +56,18 @@ public sealed class NewAzureResourceCmdlet : ResourceManipulationCmdletBase
...
         /// <summary>  
-        /// Gets or sets the plan object.  
+        /// Gets or sets the Sku object.  
         /// </summary>  
+        [Alias("SkuObject")]  
         [Parameter(Mandatory = false, ValueFromPipelineByPropertyName = true, HelpMessage = "A hash table which represents sku properties.")]  
         [ValidateNotNullOrEmpty]  
-        public Hashtable SkuObject { get; set; }  
-
+        public Hashtable Sku { get; set; }
```
Comment (gandhiniraj) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/NewAzureResourceCmdlet.cs]: Same as Plan and PlanObject

File: src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/NewAzureResourceCmdlet.cs (hunk 7 above)
Comment (TianoMS) [inline src/ResourceManager/Resources/Commands.ResourceManager/Cmdlets/Implementation/Resource/NewAzureResourceCmdlet.cs]: Will fix it.

**Pull request merged**